editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx

```diff
@@ -0,0 +1,307 @@
+  rect: WindowRectangle
+  directions: ElementOutsideVisibleAreaDirection[]
+}
+
+export type ElementOutsideVisibleAreaIndicator = {
+  id: string
+  path: ElementPath
+  position: WindowPoint
+  angle: number
+  cluster: number
+}
+
+export function useElementsOutsideVisibleArea(
+  ref: React.MutableRefObject<HTMLDivElement | null>,
+  localHighlightedViews: ElementPath[],
+  localSelectedViews: ElementPath[],
+): ElementOutsideVisibleAreaIndicator[] {
+  const bounds = ref.current?.getBoundingClientRect() ?? null
+  const canvasToolbar = document.getElementById(CanvasToolbarId)?.getBoundingClientRect() ?? null
+
+  const storeRef = useRefEditorState((store) => ({
+    jsxMetadata: store.editor.jsxMetadata,
+  }))
+
+  const canvasScale = useEditorState(
+    Substores.canvasOffset,
+    (store) => store.editor.canvas.scale,
+    'useElementsOutsideVisibleArea canvasScale',
+  )
+  const canvasOffset = useEditorState(
+    Substores.canvasOffset,
+    (store) => store.editor.canvas.roundedCanvasOffset,
+    'useElementsOutsideVisibleArea canvasOffset',
+  )
+  const navigatorWidth = useEditorState(
+    Substores.restOfEditor,
+    (store) => (store.editor.navigator.minimised ? 0 : LeftPaneDefaultWidth),
+    'useElementsOutsideVisibleArea navigatorMinimised',
+  )
+
+  const elements = React.useMemo(() => {
+    return uniqBy([...localSelectedViews, ...localHighlightedViews], EP.pathsEqual)
+  }, [localSelectedViews, localHighlightedViews])
+
+  const framesByPathString = React.useMemo(() => {
+    const frames: { [key: string]: CanvasRectangle } = {}
+    for (const path of elements) {
+      const metadata = MetadataUtils.findElementByElementPath(storeRef.current.jsxMetadata, path)
+      if (
+        metadata != null &&
+        metadata.globalFrame != null &&
+        isFiniteRectangle(metadata.globalFrame)
+      ) {
+        frames[EP.toString(path)] = metadata.globalFrame
+      }
+    }
+    return frames
+  }, [storeRef, elements])
+
+  const scaledCanvasArea = React.useMemo(() => {
+    if (bounds == null) {
+      return null
+    }
+    const scaleRatio = canvasScale > 1 ? canvasScale : 1
+    return {
+      x: bounds.x * scaleRatio,
+      y: bounds.y * scaleRatio,
+      width: bounds.width * scaleRatio - navigatorWidth,
+      height: bounds.height * scaleRatio,
+    } as WindowRectangle
+  }, [bounds, navigatorWidth, canvasScale])
+
+  const scaledCanvasAreaCenter = React.useMemo(() => {
+    if (scaledCanvasArea == null) {
+      return null
+    }
+    return getRectCenter(scaledCanvasArea)
+  }, [scaledCanvasArea])
+
+  const elementsOutsideVisibleArea = React.useMemo(() => {
+    return mapDropNulls((path: ElementPath): ElementOutsideVisibleArea | null => {
+      if (scaledCanvasArea == null) {
+        return null
+      }
+      const frame = framesByPathString[EP.toString(path)]
+      if (frame == null) {
+        return null
+      }
+
+      const topLeftSkew = { x: -navigatorWidth, y: 0 } as WindowPoint
```

Comment:
Extremely minor - we have factories for constructing the various types of `Point` in `math-utils.ts`. It seems we don't have one for `WindowRectangle` though.

Reply:
maybe we can add the `WindowRectangle` helper in a separate PR